pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,507 @@
+	for i, node := range nl.Items {
+		nodeList[i] = &node
+	}
+
+	if st.args.applySoftTaints {
+		err = st.syncSoftTaints(ctx, nodeList)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{RequeueAfter: st.resyncPeriod}, nil
+}
+
+func (st *softTainter) syncSoftTaints(ctx context.Context, nodes []*corev1.Node) error {
+	nodesMap := make(map[string]*corev1.Node, len(nodes))
+
+	nu := st.nodeUtilizationFactory(st.promClient, st.promQuery)
+	nodeUsages, err := nu.NodeUsageFromPrometheusMetrics(ctx)
+	if err != nil {
+		return err
+	}
+
+	// TODO: do we need this???
+	capacities := referencedResourceListForNodesCapacity(nodes)
+
+	for _, node := range nodes {
+		if _, exists := nodeUsages[node.Name]; !exists {
+			return fmt.Errorf("unable to find metric entry for %v, query: %v", node.Name, st.promQuery)
+		}
+		nodesMap[node.Name] = node
+	}
+
+	// usage, by default, is exposed in absolute values. we need to normalize
+	// them (convert them to percentages) to be able to compare them with the
+	// user provided thresholds. thresholds are already provided in percentage
+	// in the <0; 100> interval.
+	var usage map[string]api.ResourceThresholds
+	var thresholds map[string][]api.ResourceThresholds
+	if st.args.useDeviationThresholds {
+		// here the thresholds provided by the user represent
+		// deviations from the average so we need to treat them
+		// differently. when calculating the average we only
+		// need to consider the resources for which the user
+		// has provided thresholds.
+		usage, thresholds = assessNodesUsagesAndRelativeThresholds(
+			filterResourceNames(nodeUsages, st.resourceNames),
+			capacities,
+			st.args.thresholds,
+			st.args.targetThresholds,
+		)
+	} else {
+		usage, thresholds = assessNodesUsagesAndStaticThresholds(
+			nodeUsages,
+			capacities,
+			st.args.thresholds,
+			st.args.targetThresholds,
+		)
+	}
+
+	// classify nodes in under and over utilized. we will later try to move
+	// pods from the overutilized nodes to the underutilized ones.
+	nodeGroups := classifier.Classify(
+		usage, thresholds,
+		// underutilization criteria processing. nodes that are
+		// underutilized but aren't schedulable are ignored.
+		func(nodeName string, usage, threshold api.ResourceThresholds) bool {
+			if nodeutil.IsNodeUnschedulable(nodesMap[nodeName]) {
+				log.V(2).Info(
+					"Node is unschedulable, thus not considered as underutilized",
+					"node", nodesMap[nodeName],
+				)
+				return false
+			}
+			return isNodeBelowThreshold(usage, threshold)
+		},
+		// overutilization criteria evaluation.
+		func(nodeName string, usage, threshold api.ResourceThresholds) bool {
+			return isNodeAboveThreshold(usage, threshold)
+		},
+	)
+	var lowNodes []*corev1.Node
+	var apprNodes []*corev1.Node
+	var highNodes []*corev1.Node
+	categories := []string{"underutilized", "overutilized"}
+
+	classifiedNodes := map[string]bool{}
+	for i := range nodeGroups {
```

Comment:
`lowNodes`, `highNodes` and `apprNodes` slices may all be populated inside the classifiers. To classify the `apprNodes` you only need a third classifier that returns always `true`. I think this will simplify a lot here.

Reply:
I was thinking about that.
But I also see an additional case that I'm not sure we can solve in that way.
Now the SoftTainter is removing all the softtaints from under utilized nodes, adding one to appropriately utilized and two to overutilized.
When we have dynamic thresholds and the system is converging, all the nodes are ideally expected to be appropriately utilized. But in that case we do not want to set any soft taint.
That logic is actually here and i'm not sure I can replicate it with an additional classifier.

Reply:
Do you mean doing something like this ? Just checking if I understood.

```go
if len(low) == 0 {
    low = appropriate
    appropriate = []
}
```

Reply:
I tried adding the third, catch all, classifier at the end of the list, but i'm not sure it's ever called.